kernel/shared/src/main/scala/cats/effect/kernel/GenConcurrent.scala

```
@@ -77,7 +89,7 @@ trait GenConcurrent[F[_], E] extends GenSpawn[F, E] {
               case st @ Done(value) =>
                 st -> F.pure(value)
             }.flatten
-          }.rethrow
+          }.flatMap(_.embed(F.canceled *> F.never))
```

I adopted prior art from `GenSpawn#race` where if the memoized effect results in cancellation and the attempt to get the memoized value is masked, the result is non-termination. Though, it'd seem more user-friendly to fail fast with an error in both cases. I'm happy to switch to that behavior unless there are objections.

---

I do agree it's more user-friendly to fail fast with an error, but given that we don't know `E`, I'm not sure how we could do that here. :-( This is an unfortunate hole in cancelation right now that I'm still thinking about. Haskell solves this problem by having `forkIO` inherit mask status, so we could just add a `startWithMasks` or something like that to try to resolve that problem, but it feels very ad hoc.

---

Ah yes, still getting used to the parametric error type 😄